libV2/config/index.js

```diff
@@ -0,0 +1,222 @@
+
+const { truthy, envNamespace } = require('../constants');
+const configSchema = require('./schema');
+
+const _typeCasts = {
+    bool: val => truthy.has(val.toLowerCase()),
+    int: val => parseInt(val, 10),
+    list: val => val.split(',').map(v => v.trim()),
+};
+
+
+function _definedInEnv(key) {
+    return process.env[`${envNamespace}_${key}`] !== undefined;
+}
+
+function _loadFromEnv(key, defaultValue, type) {
+    const envKey = `${envNamespace}_${key}`;
+    const value = process.env[envKey];
+    if (value !== undefined) {
+        if (type !== undefined) {
+            return type(value);
+        }
+        return value;
+    }
+    return defaultValue;
+}
+
+const defaultConfigPath = path.join(__dirname, '../../config.json');
+
+class Config {
+    /**
+     * Returns a new Config instance merging the loaded config with the provided values.
+     * Passed values override loaded ones recursively.
+     *
+     * @param {object} overrides - an object using the same structure as the config file
+     * @returns {Config} - New Config instance
+     */
+    constructor(overrides) {
+        this._basePath = path.join(__dirname, '../../');
+        this._configPath = _loadFromEnv('CONFIG_FILE', defaultConfigPath);
+        this._defaultsPath = path.join(__dirname, 'defaults.json');
+
+        this.host = undefined;
+        this.port = undefined;
+
+        this.healthChecks = undefined;
+        this.logging = { level: 'debug', dumpLevel: 'error' };
+
+        this.redis = undefined;
+        this.warp10 = undefined;
+
+        // read config automatically
+        const loadedConfig = this._loadConfig();
+        let parsedConfig = Config._parseConfig(loadedConfig);
+        if (overrides && typeof overrides === 'object') {
+            parsedConfig = this._recursiveUpdate(parsedConfig, overrides);
+        }
+        Object.assign(this, parsedConfig);
+    }
+
+    static _readFile(path) {
+        try {
+            const data = fs.readFileSync(path, {
+                encoding: 'utf-8',
+            });
+            return JSON.parse(data);
+        } catch (error) {
+            // eslint-disable-next-line no-console
+            console.error(`error reading config file at ${path}`, { error });
+            throw new Error(`could not parse config file: ${error.message}`);
+        }
+    }
+
+    _loadDefaults() {
+        return Config._readFile(this._defaultsPath);
+    }
+
+    _loadUserConfig() {
+        return Joi.attempt(
+            Config._readFile(this._configPath),
+            configSchema,
+            'invalid Utapi config',
+        );
+    }
+
+    _recursiveUpdateArray(parent, child) {
+        const ret = [];
+        for (let i = 0; i < Math.max(parent.length, child.length); i += 1) {
+            ret[i] = this._recursiveUpdate(parent[i], child[i]);
+        }
+        return ret;
+    }
+
+    _recursiveUpdateObject(parent, child) {
+        return Object.keys(parent)
+            .concat(Object.keys(child))
```

Comment:
You can probably replace the `concat` and first `reduce` by constructing a `Set` of keys.